File: chapter1/problem1_1/problem1_1.cpp
#include <iostream>
#include <string>
using namespace std;

int isUnique(string word){
    int seen_characters[256];

    // Setting all seen characters to false
    for(size_t i = 0; i < (sizeof(seen_characters) / sizeof(seen_characters[0])); i++){
        seen_characters[i] = 0;
    }


    for(int i = 0; i < word.length(); i++){
        if(seen_characters[word.at(i)]){
            return 0;
        }
        seen_characters[word.at(i)] = 1;
    }
    return 1;
}

int main(int argc, char* argv[]) {

    if(isUnique(argv[1])){
        cout << "True" << '\n';
    } else {
        cout << "False" << '\n';
    }

    return 0;


}

File: chapter1/problem1_1/problem1_1.py
import sys

def IsUnique1(word):
    recorded_characters = []

    for character in word:
        if character in recorded_characters: 
            return False
        recorded_characters.append(character)
    return True
    

def IsUnique2(word):
    seen_characters = [False] * 256
    for character in word:
        if seen_characters[ord(character)]:
            return False
        seen_characters[ord(character)] = True
    return True



def main():
    '''Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you
    cannot use additional data structures?
    Hints: #44, #117, #132
    '''
    
    print(IsUnique1(sys.argv[1]))
    print(IsUnique2(sys.argv[1]))





if __name__ == "__main__":
    main()



